src/math.rs

```diff
@@ -361,11 +339,175 @@ impl Quat {
+        assert_eq!(c, Quat::new(2.0, 1.0, 1.0, 2.0));
+    }
+
+    #[test]
+    fn subtract() {
+        let a: Vec3 = Vec3::new(1.0, 3.0, 1.0);
+        let b: Vec3 = Vec3::new(0.5, 1.2, 1.5);
+        let c: Vec3 = a - b;
+        assert_eq!(c, Vec3::new(0.5, 1.8, -0.5));
+    }
+
+    #[test]
+    fn neg() {
+        let a: Vec3 = Vec3::new(1.0, 2.0, 3.0);
+        let b = -a;
+        assert_eq!(b, Vec3::new(-1.0, -2.0, -3.0));
+    }
+
+    #[test]
+    fn mul_f32() {
+        let a: Vec3 = Vec3::new(1.0, 2.0, 3.0);
+        let b: f32 = 10.0;
+        let c: Vec3 = a * b;
+        assert_eq!(c, Vec3::new(10.0, 20.0, 30.0));
+    }
+
+    #[test]
+    fn mul_quat() {
+        let a: Vec3 = Vec3::new(1.0, 2.0, 3.0);
+        let b: Quat = Quat::new(1.0, 0.5, 2.0, 1.5);
+        let c: Quat = a * b;
+        assert_eq!(c, Quat::new(-9.0, -2.0, 2.0, 4.0));
+    }
+
+    #[test]
+    fn sq_mag() {
+        let a: Vec3 = Vec3::new(1.0, 2.0, 3.0);
+        let b: f32 = a.sq_mag();
+        assert_eq!(b, 14.0);
+    }
+
+    #[test]
+    fn mag() {
+        let a: Vec3 = Vec3::new(1.0, 2.0, 3.0);
+        let b: f32 = a.mag();
+        assert_eq!(b, 14.0f32.sqrt());
+    }
+
+    #[test]
+    fn rotate() {
+        let a: Vec3 = Vec3::new(1.0, 0.0, 0.0);
+        let b: Quat = Quat::new(3.0, 2.0, 1.0, 1.0);
+        let c: Vec3 = a.rotate(b);
+        assert_eq!(c, Vec3::new(11.0, 10.0, -2.0));
+    }
+}
+
+#[cfg(test)]
+mod quat_tests {
+    use super::*;
+
+    #[test]
+    fn new() {
+        let a: Quat = Quat {
+            r: 1.0,
+            i: 2.0,
+            j: 3.0,
+            k: 4.0,
+        };
+        let b: Quat = Quat::new(1.0, 2.0, 3.0, 4.0);
+        assert_eq!(a, b);
+    }
+
+    #[test]
+    fn add_quat() {
+        let a: Quat = Quat::new(1.0, 1.0, 1.0, 1.0);
+        let b: Quat = Quat::new(0.5, 1.0, 0.5, 1.0);
+        let c: Quat = a + b;
+        assert_eq!(c, Quat::new(1.5, 2.0, 1.5, 2.0));
+    }
+
+    #[test]
+    fn add_f32() {
+        let a: Quat = Quat::new(1.0, 0.5, 1.0, 2.0);
+        let b: f32 = 10.0;
+        let c: Quat = a + b;
+        assert_eq!(c, Quat::new(11.0, 0.5, 1.0, 2.0));
+    }
+
+    #[test]
+    fn sub() {
+        let a: Quat = Quat::new(1.0, 1.0, 1.0, 1.0);
+        let b: Quat = Quat::new(0.5, 1.0, 0.5, 1.0);
+        let c: Quat = a - b;
+        assert_eq!(c, Quat::new(0.5, 0.0, 0.5, 0.0));
+    }
+
+    #[test]
+    fn neg() {
+        let a: Quat = Quat::new(1.0, 1.0, 1.0, 1.0);
+        let b: Quat = -a;
+        assert_eq!(b, Quat::new(-1.0, -1.0, -1.0, -1.0));
+    }
+
+    #[test]
+    fn mul_quat() {
+        let a: Quat = Quat::new(1.0, 1.0, 1.0, 1.0);
+        let b: Quat = Quat::new(0.5, 1.0, 0.5, 1.0);
+        let c: Quat = a * b;
+        assert_eq!(c, Quat::new(-2.0, 2.0, 1.0, 1.0));
+    }
+
+    #[test]
+    fn mul_f32() {
+        let a: Quat = Quat::new(0.5, 1.0, 0.5, 1.0);
+        let b: f32 = 2.0;
+        // Only should multiply `r` (the first number)
```

Comment:
What? Multiplying by a scalar scales the whole quaternion, which is reflected in this test.